ophyd/signal.py

```diff
@@ -1154,26 +1172,35 @@ def get_setpoint(self, *, as_string=None, connection_timeout=1.0,
         connection_timeout : float, optional
             If not already connected, allow up to `connection_timeout` seconds
             for the connection to complete.
+        form : {'time', 'ctrl'}
+            PV form to request
         '''
         if as_string is None:
             as_string = self._string
 
         self.wait_for_connection(timeout=connection_timeout)
-        info = self._write_pv.get_with_metadata(as_string=as_string, **kwargs)
+        info = self._write_pv.get_with_metadata(as_string=as_string, form=form,
+                                                **kwargs)
 
         if info is None:
             timeout = kwargs.get('timeout', None)
-            raise TimeoutError(f'Failed to read {self._write_pvname} within '
+            raise TimeoutError(f'Failed to read {self._setpoint_pvname} within '
                                f'{timeout} sec')
 
         value = info.pop('value')
         if as_string:
             value = waveform_to_string(value)
 
-        # The following will update all metadata, run subscriptions, and
-        # also update self._readback such that this value can be accessed
-        # through EpicsSignal.value
-        self._write_changed(value=value, **info)
+        has_monitor = self._monitors[self.setpoint_pvname] is not None
+        if form != 'time' or not has_monitor:
```

Comment:
Why does the monitor enter into this logic?  I understand updating this if it ctrl from, but why update it if the form in 'time' if there is not a monitor?

Reply:
The comment below touches on it:
```
# A different form (such as 'ctrl') holds additional data not available
# through the DBR_TIME channel access monitor
```
Part 1 of the `if form != 'time'`: It may include an updated 'units' or something else not in the dbr_time request. So, update the metadata because the next monitor event will not include it.

Part 2 of the `if not has_monitor`: means that no CA monitor is setup, so metadata will never be updated by a monitor callback. This is the only place it will be updated.

If you can find a more concise way to express that, I'd be happy to implement it.

Reply:
I read that comment as only applying to part 1.  If there is a monitor, is there harm in updating the meta-data anyway?

Maybe just add the line "if there is not a monitor, this is the only chance to update the meta-data so do it here"?  It is possible that we lost a EventResponse so that this `get` has meta-data we have not heard about yet.

If the subscription is of a ctrl type (does that make sense?) are we painted in to a weird corner?

Reply:
I get your hesitation, and I'm not entirely happy with this just yet - despite it being a few iterations from my first Signal refactor PR. The main issue is that these things remain racy, so updating the metadata at the wrong moment could wipe away what came from the monitor.

I don't think we have the right level of calls to the control layer to lock this appropriately without causing deadlock. That said, I'd like to see if we can address that raciness along with atomic updates of value + metadata (as `dict.update` is atomic in Python) in follow-up PRs.